enarx-keep-sgx-shim/src/start.S

```diff
@@ -56,26 +117,112 @@ shp:                            # The bottom of the secure heap
+    savep                                           # Save untrusted preserved registers
+
+    cmp     $0,                     %rax            # If we are returning from a syscall...
+    jne     .Lsyscall                               # ... finish the job.
+
+    push    %rsp                                    # Argument for enter()
+    push    %r11                                    # Argument for enter()
+
+    zerop                                           # Clear preserved registers
+    zerot                                           # Clear temporary registers
+    zerof   %r11                                    # Clear CPU flags
+
+    # void enter(rdi, rsi, rdx, tcs, r8, r9, &aex[CSSA-1], ctx);
+    call    enter                                   # Call enter()
+    add     $16,                    %rsp            # Remove parameters from stack
+
+    # Prepare CPU context for exit
+    zerot                                           # Clear temporary registers
+    zeroa                                           # Clear argument registers
+    zerof   %r11                                    # Clear CPU flags
+    mov     $~0,                    %r11            # Indicate ERESUME to VDSO handler
+
+    # ENCLU[EEXIT]
+.Leexit:
+    loadp                                           # Load preserved registers
+    pop     %rsp                                    # Restore the untrusted stack
+    mov     $4,                     %rax
+    enclu
+
+# %rax = syscall return stack pointer
+# %rbx = next non-enclave instruction
+# %rcx = &TCS
+# %r10 = untrusted stack pointer
+# %r11 = &aex[CSSA - 1]
+# %rsp = trusted stack pointer
+.Lsyscall:
+    movq    $0,                     SRSP(%r11)      # Clear syscall return stack pointer field
+    mov     %rax,                   %rsp            # Restore the syscall return stack pointer
+    mov     %rdi,                   %rax            # Correct syscall return value register
+    loadp                                           # Restore trusted preserved registers
+    zeroa                                           # Clear the argument registers
+    zerot                                           # Clear the temporary registers
+    zerof   %r11                                    # Clear CPU flags
+    ret                                             # Jump to address on the stack
+
+    # int syscall(rdi, rsi, rdx, aex, r8, r9, r10, rax, ctx);
+    .text
+    .globl syscall
+    .type syscall, @function
+syscall:
+    savep                                           # Save preserved registers
+    mov     %rsp,                   SRSP(%rcx)      # Save restoration stack pointer
+
+    mov     0x38(%rsp),             %r10            # Export %r10
+    mov     0x40(%rsp),             %r11            # Export %rax in %r11
+    mov     0x48(%rsp),             %rsp            # Get the exit context
+
+    xor     %rcx,                   %rcx            # Clear %rcx
```

Comment:
It seems better to jump to some label `clear_regs_before_eexit` here to avoid having to duplicate this critical code?

Reply:
We always attempt to clear the maximum registers for the exit type. Because we have different exit types, we have different registers cleared. I'll see if there is some way to consolidate this.